 <div class="alert alert-block alert-info" style="margin-top: 20px">
 <a href="http://cocl.us/SC0101ENNotebookadd"><img src = "https://ibm.box.com/shared/static/kr8cc5o4va552yrgo83j54mishtgzj2n.png" width = 750, align = "center"></a>



<h1 align=center><font size = 5>Product Link</font></h1> 

<a href="https://www.bigdatauniversity.com"><img src = "https://ibm.box.com/shared/static/ugcqz6ohbvff804xp84y4kqnvvk3bq1g.png" width = 300, align = "center"></a>

<h1 align=center><font size = 5>Scala 101</font></h1>

# Module 3 Case Objects and Classes

# Table of content

<div class="alert alert-block alert-info" style="margin-top: 20px">
<li><a href="#ref1">Companion Objects</a>
<li><a href="#ref2">Case Classes and Case Objects</a>
<li><a href="#ref3"> Apply and Unapply </a>
<li><a href="#ref4"> Synthetic Methods </a>
<li><a href="#ref4">  Immutability in Thread Safety </a>
<p></p>
Estimated Time Needed: <strong>30 min</strong>
</div>
 

<a id="#ref1"></a>

## 3.1 Companion Objects

### By the end of this section you should be able to:

<ul>
  <li>  leverage accessibility keywords </li>
  <li>describe the role companion objects play in, Scala   </li>
  <li>outline how to use companion objects </li>
</ul>  


### Public and Private fields 

We can use keywords to limit the visibility of class fields. Consider the class "Hello" and the class  "Welcome " the classes are identical but the field "message" in the class "Hello" is set to private:  

In [1]:
class  Hello { private val message: String="Hello!" }

In [2]:
class Welcome{val message: String="Hello!" }

If you construct an object of class hello:

In [3]:
val hello=new Hello()

The field "message" is private, so you can not access it:

In [4]:
hello.message

Name: Compile Error
Message: <console>:29: error: value message in class Hello cannot be accessed in Hello
       hello.message
             ^
StackTrace: 

 When we create a new instance  of class welcome, we have access to the field 

In [121]:
val welcome=new Welcome()
welcome.message

Hello!

 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  3.1: </h1>
<p> 
Create a class:"BankAccount", the parameters should be : Name, balance and  pin number.  They should be set as unmutable fields in the class constructor. For numerical values the default values should be zero and the pin number should be private.</b>
</p>
</div>

 <div align="right">
<a href="#q1" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="q1" class="collapse">
```
class BankAccount(val Name:String, val balance:Float =0, private val PinNumber:Int=0  )
```
</div>

 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  3.2: </h1>
<p> 
Create two objects John and Sara of class "BankAccount"  the fields should be set as follows:
<p> 
<p> 
John
</p> 
<p> 
Name: John Smith
</p> 

<p> 
balance: $12500
</p> 
<p> 
PinNumber:1224 
</p> 
<p> 
Sara
</p> 
<p> 
Name: Sara Ski
</p> 
<p> 
balance:0
</p> 
<p> 
PinNumber:2222
</p> 
</p>
</div>

 <div align="right">
<a href="#q2" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="q2" class="collapse">

<p>
val John=new BankAccount(Name="John Smith", balance=12500,PinNumber=1224 )
</p>
<p>
val Sara=new BankAccount(Name="Sara Ski",PinNumber=2222 )
</p>

</div>

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  3.3: </h1>
<p> 
Acess the fields for the object John, what happens when you access the field "PinNumber".</b>
</p>
</div>

  <div align="right">
<a href="#q3" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="q3" class="collapse">
```
John.Name
John.balance
you get an error when you access the field "PinNumber"
John.PinNumber


```
</div>

## Companion Objects

 If a Singleton object and a class share the same name and are located in the same source file, they are called companions. A companion class can access fields and  methods inside of its companion object.

In [123]:
object Hello{ val defaultMessage: String="Hi" }
class Hello(){
    import Hello._


    println(Hello.defaultMessage)}


In [124]:
new Hello

Hi


Hello@345487bb

<a id="ref2"></a>
## 3.2 Case Classes and Case Objects

### By the end of this section you should be able:

<ul>
    <li>describe when to use case classes   </li>
  <li> be able to describe when to you use case classes and cases objects instead of regular classes </li>


</ul>  

 Case classes are a particular class that can be used to represent data. It is simpler to create a case class than a regular class.  In a case class,  the parameters are immutable fields by default.

We can create the case class time with the parameters "hours" and "minutes":

In [7]:
case class Time(hours: Int=0, minutes:Int=0)

We can create a new instance of that class and assign it to a value:

In [8]:
val time1=Time(9,10)

We can represent the instance or object "time1"  with a table,  the field names are in the right column, and the field values are located in the left column.

 <img src= "https://ibm.box.com/shared/static/efawhc7immzeuopfnaskio4pj9iogt8j.png" width = 400></a>

We can view the actual fields 

In [9]:
time1.hours

9

In [10]:
 time1.minutes

10

 If we display the object, we get the class type along with the values the fields are set to

In [11]:
time1

Time(9,10)

We can create a second instance of "Time" called "time2",  we set the field values to be the same.

In [12]:
val time2=Time(9,10)

Similarly, we can visualize the object using a table 

<img src = "https://ibm.box.com/shared/static/q3qlt4en8s8jk2tg1i7z6nhsxwbsjxvn.png" width = 400>

 We can view the actual fields

In [13]:
time2.hours

9

In [14]:
time2.minutes

10

 <a id="ex1"></a>we can also do equivalence checks seeing if two instances of the same class are the same. If the fields are the same, we get true.

In [15]:
time1==time2

true

 The procedure is summarized in the figure below, where each table is compared.  As both fields are  the same we get true :

<img src = "https://ibm.box.com/shared/static/0mzmiud464w7vqxjuwn1smessqbgh4ud.png" width = 1000>

we can create a new instance without using the keyword:

In [16]:
val time3=Time(9,20)

If we perform an equivalence check we get a false as the fields are different 

In [17]:
time2==time3

false

We can see that the field minutes in "time3" is set to 20 , hence the equivalence operation produces a false, as shown in the following figure: 

<img src = "https://ibm.box.com/shared/static/ge6706g10yp62u28k6p7xjftoqfsjgzd.png" width = 1000>

You cannot  create a case class without parameters 

In [18]:
case class Data

Name: Compile Error
Message: <console>:1: error: case classes without a parameter list are not allowed;
use either case objects or case classes with an explicit `()' as a parameter list.
case class Data
               ^
StackTrace: 

 but you can create a case object 

In [19]:
case object Data

There is no need to instantiate it 

In [9]:
Data

Data

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  3.4: </h1>
<p> 
Create a case class object Birthday that represents the birthdate of a person. With the integer fields in the following order: day, the month (from 1 to 12) and year.  .</b>
</p>
</div>

  <div align="right">
<a href="#q34" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="q34" class="collapse">
```
case class Birthday(day:Int,month:Int, year:Int ) 


```
</div>

 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  3.5: </h1>
<p> 
Create an instance of the case class Birthday called Pat, born on 11 th of October (11 month) in 1991.</b>
</p>
</div>

<div align="right">
<a href="#q35" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="q35" class="collapse">
```
val Pat=Birthday(11,10,1991)


```
</div>

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  3.6: </h1>
<p> 
Print out all the fields of the instance "Pat" along with the instance itself.</b>
</p>
</div>

 <div align="right">
<a href="#q36" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="q36" class="collapse">
```
println("day",Pat.day)
println("mounth",Pat.month)
println("year", Pat.year)
println("instance of object:",Pat)


```
</div>

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  3.7: </h1>
<p> 
Create an instance of the case class Bob, set all the fields to the exact same value as Pat, but change the value of year to 1992.  Finally, perform an equivalence check. of Bob and Pat.</b>
</p>
</div>

<div align="right">
<a href="#q37" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="q37" class="collapse">
```
val Bob=Birthday(11,10,1992)
Bob==Pat

```
</div>

<a id="ref3"></a>
## 3.4 Apply and Unapply 

### By the end of this section you should be able to:

<ul>

  <li>Illustrate the difference between a type and a term   </li>
  <li>Describe how the apply method works in both objects and classes </li>
<li>Outline how unapply works </li>
</ul>  


### Type
<li>A type is a description of a concept in an application. For example, a class is a type.</li>


### Term
<ul>
<li>A term is a concrete representation of a type </li>
<li>Any class instance, including an object, is a term </li>
<li>Methods  are also a term </li>
</ul>  


### Apply 

 When you create a case class you are actually creating a companion object  to create the instance of that class 

In [10]:
case class Time(hours: Int=0, minutes:Int=0)

we can verify this by typing the name of the class we get the object: 

In [11]:
Time

Time

 The object is a factory for the instances of the case class.The fields of the method are set using the special method apply. You do not have to call the method directly, the next two lines generate instances with the same field values 

In [12]:
Time.apply(9,0)

Time(9,0)

In [13]:
Time(9,0)

Time(9,0)

 We can see an example of the apply method by creating the object "Reverse". The method apply takes the string "s" and reverses  it: 

In [15]:
object Reverse{ def apply(s: String):String=s.reverse}

We can call the apply method explicitly 

In [17]:
Reverse.apply("hello")

olleh

or just using parentheses 

In [18]:
Reverse("hello")

olleh

 Something similar happens when you create an array   (we will cover arrays in the next section).  We can use the apply method or merely use the parentheses as shown in the following example:

In [148]:
val A=Array.apply(1,2,3)
println("A(0):",A(0))
println("A(1):",A(1))
println("A(2):",A(2))

(A(0):,1)
(A(1):,2)
(A(2):,3)


In [149]:
val B=Array(1,2,3)
println("B(0):",B(0))
println("B(1):",B(1))
println("B(2):",B(2))

(B(0):,1)
(B(1):,2)
(B(2):,3)


The following will be used for the next question. The class "OffTime" takes the first name of the employee, the employee's lunch time and break time as fields.

In [20]:
class OffTime(val Name:String, val LunchTime : String , val BreakTime:  String )

 The following companion object creates an instance of the class 

In [21]:
object OffTime {
    
    def apply(Name:String, LunchTime : String, BreakTime: String)={
        
        new OffTime(Name, LunchTime, BreakTime)
        
    }
    
}

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  3.8: </h1>
<p> Use the companion object 'OffTime' to create an instance of the class called 'Employ1' and print the object:
<p>
Name="John"
</p>
<p>
LunchTime="12:30 pm"
</p>
<p>
BreakTime="3:15 pm"
</p>
</b>
</p>
</div>

In [29]:
val Employ1=OffTime("John ","12:30 pm","3:15 pm")
 Employ1

OffTime@3dea9477

<div align="right">
<a href="#q38" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="q38" class="collapse">
<p>
val Employ1=OffTime("John ","12:30 pm","3:15 pm")

</div>

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  3.9: </h1>
<p>  Print out the filds of the instance:
</p>

</div>

Name: John 
Lunch Time: 12:30 pm
Break Time: 3:15 pm


<div align="right">
<a href="#q39" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="q39" class="collapse">

<p>
println("Name: "+Employ1.Name)
</p>
<p>
println("Lunch Time: "+Employ1.LunchTime)
</p>
<p>
println("Break Time: "+Employ1.BreakTime)
</p>
</div>

 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  3.10: </h1>
<p>  Create a case class "OffTime1" with the identical fields "OffTime" and create an instance called "Employ2" with the exact same fields as 'Employ1'and print out the instance 
</p>

</div>

<div align="right">
<a href="#q310" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="q310" class="collapse">

<p>
case class OfftimeCase(Name:String, LuchTime : String, BreakTime: String)
</p>
val Employ2=OfftimeCase("John ","12:30 pm","3:15 pm")
<p>
print(Employ2)
</p>
</div>

 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  3.11: </h1>
<p>  Print out the fields  of the instance "Employ2":
</p>

</div>

 <div align="right">
<a href="#q311" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="q311" class="collapse">

<p>
println("Name: "+Employ2.Name)
</p>
<p>
println("Luch Time: "+Employ2.LuchTime)
</p>
<p>
println("Break Time: "+Employ2.BreakTime)
</p>
</div>

### Unapply 

The method Unapply constructs the instance of the case class. For example, if I create an instance of "Time " with hours set to 8 and minutes set to 30:

In [154]:
val time1=Time(8,30)


I can use the unapply method from the companion object to get the values of the field in this case hours

In [155]:
Time.unapply(time1)

Some((8,30))

 The procedure is summarized in the following figure, the instance or object time is represented by a table. Each of the different fields is represented with different rows in the table using colors to help distinguish them. After unapply is used, the field values are mapped to a tuple.

 <img src = "https://ibm.box.com/shared/static/x4mxyvlem4sa7k8h17902w941bcj5dm7.png" width = 1200>



<a id="ref4"></a>
## Synthetic Methods

#### By the end of this section you should be able:

<ul>

  <li> describe how the Scala compiler generates functionality </li>
  <li>explain the methods: equals, hashCode, toString and copy </li>
<li> outline how you would use immutable case classes what state is changing  </li>
</ul>  


Synthetic methods are prebuilt methods that are included when you build a case class, similar to the apply method.

#### equals()

for "equals()" check out the <a href="#ex1"> previous example </a>

The method "hashCode()" maps instances with different fields to a number,consider the case class time.

In [54]:
case class Time(hours: Int=0, minutes:Int=0)

We can create an instance of the case class with the field hours set to 8 and the field minutes set to 30.

In [39]:
val time=Time(8,30)
time

Time(8,30)

 When we create the instance of the case class the compiler generates the method ".hashCode()" . We can use the method to generate a value that we can use later in a "HashMap" or "HashSet".

In [40]:
time.hashCode()

448866544

 The method maps the instance to the number '448866544'

 If we create a different instance and apply the method  "hashCode()" we get a different value.

In [53]:
 Time(8,40).hashCode()==Time(8,50).hashCode()


false

In [41]:
val time1=Time(8,40)
time1.hashCode()

-1116640459

The following image summarizes the result, after applying the method ".hashCode()" to each instance we get a different value.

 <img src = "https://ibm.box.com/shared/static/yfxums8j66verynqzecaesj2n7zafq1m.png" width = 1200>



 If we create two instances with the same fields, then apply the method ".hashCode()" the same value is generated: 

In [177]:
val time3=time.copy(8,40)
time3.hashCode()

-1116640459

In [178]:
val time4=Time(8,40)
time1.hashCode()

-1116640459

 The result is summarised in the following figure:

<img src = "https://ibm.box.com/shared/static/puo74z2fshvxltt12sxcommwrl4g46y0.png" width = 1200>

### toString()

When you generate a case class the method "toString()" is generated. Consider the class Dog with the field name and breed, when we create a new instance of the class and print out the name we see the virtual memory space.

In [42]:
class Dog(name:String,Breed:String )
print(new Dog("Max","German Shepard"))

$line112.$read$$iw$$iw$Dog@3064156b

When you create a case class the method "toString()" is generated, when you print out an instance of the case class the name of the case class and the fields are displayed. Consider the case class K9 with the field name and Breed when you print the instance of the object you get the name of the class as well as the field's value.

In [43]:
case class K9(name:String,Breed:String)
print(K9("Max","German Shepard"))

K9(Max,German Shepard)

### copy()

The method copy allows you to copy an instance of a case class; it also gives you the ability to change only a few fields. Consider the instance of the case class K9 called dog1:


In [44]:
val dog1=K9("Max","German Shepard")
dog1

K9(Max,German Shepard)

We can create a new instance and call it dog2:

In [45]:
val dog2=dog1.copy()
dog2


K9(Max,German Shepard)

we can check if the two instances are equal 

In [46]:
dog1==dog2

true

we can copy the instance to dog3 and change the field name to Bob 

In [47]:
val dog3=dog1.copy(name="Bob")
dog3

K9(Bob,German Shepard)

we can verify  the instances are not equal 

In [39]:
dog1==dog3

false

 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  3.12: </h1>
<p>  Create an instance of the case class K9,  called it "D1" and give it the name "Killer" and the breed "Chihuahua".  Use the copy method to create a new instance changing the name to Orange, call this new instance "D2":
</p>

</div>

<a href="#q312" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="q312" class="collapse">

<p>
val D1=K9("Killer" ,"Chihuahua" )
</p>
println(D1)
<p>
val D2=D1.copy(name="Orange")
</p>
println(D2)
</p>
</div>

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  3.13: </h1>
<p>Will you get the same value if you apply the method "hashCode()" ?:
</p>

</div>

(The hash code of D1:,-818883252)
(The hash code of D1:,-1527973445)



<a href="#q313" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="q313" class="collapse">



<p>
println("The hash code of D1:",D1.hashCode())
</p>
<p>
println("The hash code of D1:",D2.hashCode())
</p>
</div>

<a id="ref5"></a>
# Immutability in thread safety 

### By the end of this section you should be able to:

<ul>

  <li>  understand basic thread safety in the JVM  </li>
  <li>described the importance of immutability in multithreaded applications  </li>
<li> online how to use snapshots to preserve thread safety with case classes  </li>
</ul>  


consider the case class customer 

In [6]:
case class Customer(fistName: String,lastName:String)

we can create a new instance of the class and assign it to the "val" person1:

In [11]:
val person1 =new Customer("Joie","JoJo")
person1

Customer(Joie,JoJo)

we can not assign the value to a new instance of customer as val's are immutable 

In [26]:
person1 =new Customer("Joie","JoJo")

Name: Compile Error
Message: <console>:29: error: reassignment to val
       person1 =new Customer("Joie","JoJo")
               ^
StackTrace: 

we can access the fields 

In [27]:
person1.fistName

Joie

 but as the  are val's we can not change them 

In [29]:
person1.fistName="john"

Name: Compile Error
Message: <console>:29: error: reassignment to val
       person1.fistName="john"
                       ^
StackTrace: 

If we assign an instance of a case class to a var we must be more careful. To ensure that any threads  know about any changes we use the "@volatile"  annotation:

In [30]:
@volatile var Person2=Customer("Martin","Odersky")
Person2

Customer(Martin,Odersky)

Therefor If we use the method copy and change the field "lastname" other threads will see this change.

In [32]:
Person2=Person2.copy(lastName="Doe")
Person2

Customer(Doe,Doe)

<hr>
Copyright &copy; 2017 [cognitiveclass.ai](cognitiveclass.ai?utm_source=bducopyrightlink&utm_medium=dswb&utm_campaign=bdu). This notebook and its source code are released under the terms of the [MIT License](https://bigdatauniversity.com/mit-license/).​

# About the Authors:  

 [Joseph Santarcangelo]( https://www.linkedin.com/in/joseph-s-50398b136/) has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

